### DISCUSSION TITLE
Discussion on Katacoda removal and possible replacement

### DISCUSSION THREAD

**Discussion by reylejano** [kubernetes/website#38878] [General]
Katacoda has removed public access on June 15, 2022 ([announcement](https://www.oreilly.com/online-learning/leveraging-katacoda-technology.html))
Katacoda continued to host tutorials for kubernetes.io afte public access ceased but in December of 2022 O'Reilly communicated the intent to stop hosting tutorials for kubernetes.io.
This is GitHub discussion is about the Katacoda removal, next steps for SIG Docs, and possible replacement for Katacoda.

Kubernetes scenarios for Katacoda: https://github.com/katacoda-scenarios/kubernetes-bootcamp-scenarios

Related to issues https://github.com/kubernetes/website/issues/33936, https://github.com/kubernetes/website/issues/33918, https://github.com/kubernetes/website/issues/38785

**Comment by reylejano** [comment]
Let's state what Katacoda currently (at the time of writing this comment) provides for kubernetes.io (I'm open to suggestions and/or edits):
- Katacoda is a learning module that provided a place for folks to learn about Kubernetes with guided tutorials on kubernetes.io and Katacoda scenarios running on Katacoda's infrastructure
- Katacoda scenarios are browser-based that provided terminal access to a virtual machine where folks can access a Kubernetes cluster to learn concepts and fundamentals
- Katacoda was free and did not require a login or people to signup

**Comment by reylejano** [comment]
From Tim Bannister on the SIG Docs meeting on Jan 10, 2023: are there features that Katacoda did not offer and we should consider?

**Reply by sftim** [reply]
Our Katacoda scenarios were always in English, even in localized docs.

Example: [Привіт Minikube](https://kubernetes.io/uk/docs/tutorials/hello-minikube/) (Hello Minikube in Ukrainian)
The page is in English but clicking the Launch Terminal button produces an English-language sandbox.

**Reply by divya-mohan0209** [reply]
Towards the end, i.e., last year, afaik, performance + slowness issues were reported with the integration. Maybe not an exact dealbreaker, but the replacement should have a support system in place, as we have for Netlify, in case of slowness issues. That leads us to a separate discussion of whether we want to go down the paid/free route.

**Reply by afbjorklund** [reply]
> are there features that Katacoda did not offer and we should consider?

Katacoda offered versions of Kubernetes, that were well outside the version skew...

For the v1.26.0 release, the linked* versions of the tutorials were still using v1.20.2
\* https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/#availability

It would have been nice, if they were at least using v1.25.3 (like minikube and kind) ?

Related to issue #37817

**Comment by afbjorklund** [comment]
Killercoda offers an environment, running a pre-installed Kubernetes cluster.

* Killercoda is a learning module that provides a place for folks to learn about Kubernetes with guided tutorials on kubernetes.io and Killercoda scenarios running on Killercoda infrastructure
* Killercoda scenarios are browser-based that provides terminal (and editor!) access to a virtual machine where folks can access a Kubernetes cluster to learn concepts and fundamentals
* Killercoda is free but **does** require a login or people to signup

https://killercoda.com/kubernetes

Killercoda offers migration help, from people coming from the Katacoda platform:

* https://itnext.io/katacoda-to-killercoda-migration-guide-d21961fc0c9b

----

Sample learning environment with the cluster already running, kubectl configured.

* https://killercoda.com/examples/scenario/kubernetes-1node

Note: Killercoda does **not** offer a pre-installed minikube (or kind), only ubuntu.

* https://killercoda.com/examples/scenario/ubuntu-simple

So the "hello-minikube" or "create-cluster" scenarios, won't work out-of-the-box.
(it would be possible to document them "from scratch", if one really really wanted to)

The images come with _both_ docker and podman running, as well as the theia editor.

One bonus feature is that they offer 2 node clusters, for more advanced scenarios.
This is compensated by each node having 1 vCPU, instead of 2 vCPU like Katacoda.

Currently it is **not** possible to replicate the Killercoda node setup "at home" (offline).

**Comment by kundan2707** [comment]
one related update has been done here 
https://github.com/kubernetes/website/pull/39190

**Comment by marcosnils** [comment]
:wave: play-with-kubernetes (https://labs.play-with-k8s.com/) maintainer here. I few months ago (https://github.com/kubernetes/website/issues/34007) I brought the same topic to this group when the initial Katakoda announcement was made and the community seemed interested in evaluating the alternative of using an "open source" proved (it's already in the kubernetes docs) solution. The main reason why we decided not to move forward was because Katakoda / O'Reilly decided to keep supporting the kubernetes tutorials. Now that has come to an end.

I'm happy to resume those conversations if there's interest in bringing this topic back to the table. I'm really hoping for an open source alternative to be chosen this time so this doesn't happen again :pray:

**Reply by afbjorklund** [reply]
Looks promising! Are you planning to add more tools like "start.sh", to make the installation less DIY (with raw `kubeadm`) ?

```
You can bootstrap a cluster as follows:

 1. Initializes cluster master node:

 kubeadm init --apiserver-advertise-address $(hostname -i) --pod-network-cidr 10.5.0.0/16
    

 2. Initialize cluster networking:

kubectl apply -f https://raw.githubusercontent.com/cloudnativelabs/kube-router/master/daemonset/kubeadm-kuberouter.yaml


 3. (Optional) Create an nginx deployment:

 kubectl apply -f https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/application/nginx-app.yaml

```

For instance, it seems that Docker is not running and that `cri-dockerd` is not installed (for k8s 1.24 and beyond).

`Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?`

----

Or if there is another option, to have the cluster already up and running, that would be great when getting beyond "Step 0"

Being able to run the regular `kind` commands would also be good, then the user could follow along at home...

**Reply by marcosnils** [reply]
Yeah, we can adapt the workflows as needed. As I said before, all the code is open source and can be found here https://github.com/play-with-docker/play-with-docker. You can also run it in your local machine or set it up in your internal organization if needed.

If the SIG decides PWK could be the best thing as a solid base for future interactive docs and the kubernetes project generally, we'll need to fine tune the details about improvements needed to be made (like upgrading versions, improving start scripts, etc) and maintenance / operational expenses.

**Reply by afbjorklund** [reply]
Whatever they decide, it would be nice if the k8s installation wasn't _too_ much more than a `minikube start` or a start.sh

And if the installation scripts and the platform was Open Source, although that hasn't stopped the use of Docker Desktop in docs

**Reply by marcosnils** [reply]
> Whatever they decide, it would be nice if the k8s installation wasn't _too_ much more than a `minikube start` or a start.sh

definitely! always looking for improvement suggestions :muscle: 

> And if the installation scripts and the platform was Open Source, although that hasn't stopped the use of Docker Desktop in docs

True. However, in this particular case, we already have the unfortunate experience of Katakoda / O'Reilly deciding to throw a big part of community work (since tutorials contributions generally come from the community) away and hurting the project. I remember bringing this same exact concern when Katakoda was initially selected as a tutorials platform. We'll see how this moves forward now :pray:

**Reply by afbjorklund** [reply]
For what is worth, the runtime worked fine after upgrading and adding the missing software and crictl configuration.

The main thing was that the systemd units didn't work with the fake systemd replacement, but starting manually worked.

```
[node1 ~]$ crictl version
Version:  0.1.0
RuntimeName:  docker
RuntimeVersion:  20.10.1
RuntimeApiVersion:  v1
```

The cluster didn't start though, I guess the `kubeadm` wrapper needs some upgrade from the ancient version (1.20.1) ?

E0215 18:02:45.861748    7171 run.go:74] "command failed" err="failed to parse kubelet flag: unknown flag: --network-plugin"

/etc/systemd/system/kubelet.env:KUBELET_NETWORK_ARGS="--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"

**Reply by marcosnils** [reply]
> The cluster didn't start though, I guess the `kubeadm` wrapper needs some upgrade from the ancient version (1.20.1) ?

yeah.. probably. As highlighted above, it needs some caring since I've been mostly maintaining this ad honorem in my spare time and I had a baby 6mo ago :baby: :tongue:

**Reply by afbjorklund** [reply]
Just the usual upgrade woes, adding CRI and CNI...  `Your Kubernetes control-plane has initialized successfully!`

```console
[root@node1 ~]# kubectl get nodes -o wide
NAME    STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                KERNEL-VERSION      CONTAINER-RUNTIME
node1   Ready    control-plane   27s   v1.26.1   192.168.0.8   <none>        CentOS Linux 7 (Core)   4.4.0-210-generic   docker://20.10.1
[root@node1 ~]# kubectl get pods -A -o wide
NAMESPACE     NAME                            READY   STATUS    RESTARTS   AGE   IP            NODE    NOMINATED NODE   READINESS GATES
kube-system   coredns-787d4945fb-mzclv        1/1     Running   0          17s   10.10.0.3     node1   <none>           <none>
kube-system   coredns-787d4945fb-xkktc        1/1     Running   0          17s   10.10.0.2     node1   <none>           <none>
kube-system   etcd-node1                      1/1     Running   0          28s   192.168.0.8   node1   <none>           <none>
kube-system   kube-apiserver-node1            1/1     Running   0          27s   192.168.0.8   node1   <none>           <none>
kube-system   kube-controller-manager-node1   1/1     Running   0          26s   192.168.0.8   node1   <none>           <none>
kube-system   kube-proxy-lfpgb                1/1     Running   0          17s   192.168.0.8   node1   <none>           <none>
kube-system   kube-scheduler-node1            1/1     Running   0          26s   192.168.0.8   node1   <none>           <none>
```

**Reply by marcosnils** [reply]
@afbjorklund happy to get contributions since you're spending some cycles on this. Here's the Docekrfile that currently builds the k8s image: https://github.com/play-with-docker/play-with-docker/blob/master/dockerfiles/k8s/Dockerfile

we can remove the `docker` dependencies from there and bump the k8s version if you can spare the time :muscle:

**Reply by afbjorklund** [reply]
Or we continue to use `minikube start`, like before ?

Would be nice if there was a new [get.k8s.io](https://get.k8s.io/) script...

**Reply by marcosnils** [reply]
But with minikube users won't be able to run things in the browser? Having a tool like katakoda / killerkoda / PWK adds a lot of value for people that can't / doesn't want to spend the time to set things up and/or have very poor connectivity. Not to mention the benefit of having interactive tutorials like these ones: https://training.play-with-docker.com/multi-stage/

I think you're comparing apples to oranges by suggesting that minikube, dind or any other local cluster setup tool is equivalent to these solutions.

**Reply by afbjorklund** [reply]
The katacoda scenarios are currently running minikube (with the "none" driver), which is why I suggested it.

![katacoda-minikube](https://user-images.githubusercontent.com/10364051/219130482-23532257-5190-4612-9c8c-df714ee2a221.png)

https://kubernetes.io/docs/tutorials/kubernetes-basics/create-cluster/cluster-interactive/

* https://github.com/kubernetes/minikube/issues/15651

**Reply by marcosnils** [reply]
oh. you mean running `minikube` inside PWK? Well.. that's possible as well. :open_hands:. It might require a bit more effort but it can definitely be done. Having said that, I guess it'll be up to this sig group to decide what the suggested approach to create a cluster is.

**Reply by afbjorklund** [reply]
My suggestion is to have **two** versions of the "image", one for doing the cluster installation and one "prepared earlier"...

* So that in Step 0 "_Hello Minikube_", you can run `minikube start`. Currently that requires the container runtime installed.

* But then in Step 1 "_~~Create a~~ Connect to Cluster_", you can run `kubectl`. The local cluster is already running and configured.

Then the fledgling Ops person can watch the (30 sec) installation process but the person leaning towards Dev can dive right in:

![kubernetes-tutorials](https://user-images.githubusercontent.com/10364051/219297186-4c00cf42-a8b6-4ac4-bd4d-6769f2d77cee.png)

Possibly split the first step into three [tools](https://kubernetes.io/docs/tasks/tools/).

**Reply by marcosnils** [reply]
> * But then in Step 1 "_~Create a~ Connect to Cluster_", you can run `kubectl`. The local cluster is already running and configured.

this is a bit harder to do since bootstrapping the cluster **needs** to happen when the user starts the session. You can't have an image with a cluster already running. We can start the bootstrapping automatically as soon as the user starts the session and show him an output about the progress.

**Reply by afbjorklund** [reply]
> You can't have an image with a cluster already running

It was a feature of the killercoda images, which uses VMs.

I suppose it could also happen, if using some cloud provider for the basic tutorials. And they set up the k8s cluster for you ?
For minikube there is the "pause" feature, but that would only work the _second_ time - the first time it needs bootstrapping.

**Reply by marcosnils** [reply]
> It was a feature of the killercoda images, which uses VMs.

right, PWK uses containes since multiple user sessions can run in the same machine which brings costs down considerably. I'm aware that katakoda and killerkoda have a small pool of "warm" VMs, but very soon you hit that limit and they make you wait until the new VM is provisioned which I think it hurts the DX more.

**Reply by afbjorklund** [reply]
What is DX ?

**Reply by marcosnils** [reply]
Developer Experience. Having a loading spinner saying "Your environment will be ready shortly" is not as friendly as seeing in the actual terminal how the cluster is being bootstrapped.

**Reply by afbjorklund** [reply]
I am a bit concerned about using legacy distributions like CentOS 7 and legacy platforms like Docker Swarm for this.

**Comment by marcosnils** [comment]
@reylejano @sftim any updates about this? Is there any interest from the docs SIG or the k8s project to provide a replacement for katadacoda? I'd personally love to see something, and I currently have bandwidth and willingness to help here.

**Reply by afbjorklund** [reply]
As far as I know, the Kubernetes 1.20 online tutorials will be **shut down** in three weeks...
https://kubernetes.io/blog/2023/02/14/kubernetes-katacoda-tutorials-stop-from-2023-03-31/

Possibly the contents will be "rescued", to avoid broken Katacoda links and for offline usage.
e.g. https://github.com/katacoda-scenarios/kubernetes-bootcamp-scenarios/blob/master/1/


```
Let’s view the cluster details. We’ll do that by running *kubectl cluster-info*:

`kubectl cluster-info`{{execute}}

During this tutorial, we’ll be focusing on the command line for deploying and exploring our application.
To view the nodes in the cluster, run the *kubectl get nodes* command:

`kubectl get nodes`{{execute}}

This command shows all nodes that can be used to host our applications. Now we have only one node, and we can see that its status is ready (it is ready to accept applications for deployment). 
```

**Reply by afbjorklund** [reply]
So far, this path is the closest thing to keeping the scenarios going:
https://killercoda.com/kubernetes-bootcamp-scenarios/scenario/1

![killercoda-tutorials](https://user-images.githubusercontent.com/10364051/224701146-95be65af-7da8-4861-9710-cb9b2bbc8bb9.png)

But it does not use `minikube` and the k8s installer is not open source.
So the user would have to set up their own cluster, for off-line usage...

**Reply by marcosnils** [reply]
My main concern about using killerkoda is basically repeating the same mistakes that katakoda. We never know when / if the company will keep running and from one day to the other, all effort that the community put into creating tutorials and generating content, will be basically thrown over the board.

**Reply by afbjorklund** [reply]
I think it would be nice to have an option to run the scenarios locally (offline), as well.
Currently I can run the minikube dashboard, but there are no interactive scenarios...

More traditional copy/paste to terminal:
https://minikube.sigs.k8s.io/docs/start/

**Reply by afbjorklund** [reply]
There are some browser proof-of-concepts, but nothing that is ready for deployment yet.
It needs to have the integrated authentication, and also the dynamic scenario interaction.

![browser-minikube](https://user-images.githubusercontent.com/10364051/224707580-7d42850a-59d1-476b-956e-b30cd6f77603.png)

And there are a lot of UI bugs in minikube, to fix in the interaction with the "none" driver.
Most of them are already in the existing version (1.20), but it got even worse since 1.24...

**Comment by afbjorklund** [comment]


**Reply by sftim** [reply]
This discussion is about:
- next steps for SIG Docs given the removal of Katacoda
- possible replacement for Katacoda

Related work such as updates to those tutorials would be _in scope_ IMO. Please try not to veer too far off topic!

**Reply by afbjorklund** [reply]
Just wondered if the content is too out of date anyway, so that is not worth just moving it to a different hosting.

**Comment by letrthang** [comment]
got backup of the tutorial here :)) :
https://pwittrock.github.io/docs/tutorials/

**Comment by darshandkd** [comment]
Apologies if this is not the right channel for my ask in advnace.

We're planning to host an offline k8s lab environment for students to test their understanding of k8s training, and for the same we're looking for offline setup to host browser based k8s scenarios where students can practice/test their understanding.

Is there any offline version of Killerkoda-like systems? If yes, would someone please help me pointing the direction?

Thank you

**Reply by afbjorklund** [reply]
There is no available off-line Katacoda solution as far as I know, and the Killercoda scripts are **not** Open Source...

But it _would_ be interesting to have one available, we have multiple solutions for running Kubernetes but they are all based on a Terminal (and not a Browser) - and in their default setup require a local hypervisor instead of a local server:

1) Vagrant
2) Minikube
3) Lima

There are some existing proof-of-concept implementations, but they lack a bit in both convenience and security...

It is out of scope for _this_ issue, but it (and a free CC-BY-4.0 course) would be a nice addition to the k8s ecosystem

**Reply by marcosnils** [reply]
Play-with-kubernetes is open source and can be installed locally offline

**Reply by afbjorklund** [reply]
But it is not possible to run the existing katacoda scenarios with PWK, right ?

**Reply by marcosnils** [reply]
op didn't mention that compatibility with katacoda scenarios is required. it's asking for a similar offline based environment. PWK can do that.

**Reply by afbjorklund** [reply]
Anyway, it was considered as off topic to _this_ SIG Docs issue about removal/replacement: https://github.com/kubernetes/website/discussions/38878#discussioncomment-5399331

**Reply by darshandkd** [reply]
Thank you @marcosnils that indeed helped me with the direction. Yea, katakoda scenario isn’t a prerequisite, as we’d like to load our own scenarios.

**Reply by darshandkd** [reply]
Btw, I also came across Rancher’s HobbyFarm which seems a promising project for the objective in question.

Thank you everyone for your valuable input!

**Reply by marcosnils** [reply]
> Thank you @marcosnils that indeed helped me with the direction. Yea, katakoda scenario isn’t a prerequisite, as we’d like to load our own scenarios.

Sure np! Hobbyfarm is a great project indeed but AFAIK it doesn't have the ability to run 100% offline since it requires provisioning on-demand vm's for the environment.

**Reply by pablochacin** [reply]
@marcosnils Maybe I'm missing something, but it seems that HobbyFarm's [backend](https://github.com/hobbyfarm/gargantua) uses a terraform operator for provisioning the VMs, so un principle it should be possible to spin local VMs.

**Reply by darshandkd** [reply]
> > Thank you @marcosnils that indeed helped me with the direction. Yea, katakoda scenario isn’t a prerequisite, as we’d like to load our own scenarios.
> 
> Sure np! Hobbyfarm is a great project indeed but AFAIK it doesn't have the ability to run 100% offline since it requires provisioning on-demand vm's for the environment.

I'm yet to complete my lab setup with HobbyFarm and so far it doesn't seem straight forward, but it appears to be a complete offline setup. I'll share more once I learn more about it down the line. Thank you for the input though.

**Comment by skunkworksdev** [comment]
Hello everyone,
I do not know if this is according to your requirements. Sorry in advance if it isn't!
I have a software with the following features:
- Can run fully offline.
- Produce an iframed web terminal.
- Execute commands in the external(parent) page on the terminal, similar to Katacoda with a button click.
- Also retrieve the command execution status(success or failure) and execution time and display it on the parent page.
- If the command status and time features aren't required then, with very little modifications, the execute on click feature works successfully for remote machines with Windows, Linux etc. Otherwise only Linux is supported.
- I was also able to integrate the terminal in a react frontend.
- I am attaching a link to a YouTube video that demonstrates the software: [link](https://youtu.be/IEIJk_4zIFs). (Watch from 5 min onwards to see the execute on click part).

**Comment by sftim** [comment]
I wonder if we could use something like https://webvm.io/ (which emulates x86 purely client side)

**Reply by marcosnils** [reply]
I think making containers work in something like webvm.io will be extremely slow since all the kernel primitives for container management (cgroups, overlay fs, namespaces, etc) will have to be emulated which IMO will have a big performance penalty and probably contains some bugs.

**Reply by marcosnils** [reply]
I've offered the docs community to use https://github.com/play-with-docker/play-with-docker the **only** open sourced MIT licensed project that provides a similar katacoda experience. The way I see it, there's no interest in the community currently to bring back interactive learning tutorials.

**Comment by antonioberben** [comment]
Have you considered using educates (https://github.com/vmware-tanzu-labs/educates-training-platform) which is a cloud-native training platform that checks all the requirements? 

It's been developed for some years by VMware and just made open source. It's experience is like katacoda/killercoda, but much more feature-rich as it can run on a cluster giving the user only access to a namespace or a vcluster or even a VM, which reduces the cost of execution. 

Check some sccreenshots (https://docs.educates.dev/project-details/sample-screenshots) or the docs at (https://docs.educates.dev)

**Reply by sftim** [reply]
We are short of volunteers to deploy and manage new infrastructure here. It's an option but Kubernetes contributors with capacity to look after the project's infrastructure are a scarce resource.

**Reply by marcosnils** [reply]
That's where the foundation can actually help by providing some budget for this? Katakoda had a budget allocated and they paid the service for quite some time.

**Reply by sftim** [reply]
Is there a service we can buy? If so, which should it be?

**Reply by marcosnils** [reply]
yep, I'd suggest checking out https://instruqt.com/. That's what Hashicorp interactive tutorials use: https://learn.hashicorp.com/

**Reply by sftim** [reply]
See https://github.com/kubernetes/website/discussions/38878#discussioncomment-7060932 for the Instruqt suggestion.

**Comment by sftim** [comment]
Copying from https://github.com/kubernetes/website/discussions/38878#discussioncomment-7060716
Suggestion by @marcosnils: use https://instruqt.com/

**Comment by sftim** [comment]
We have removed Katacoda and didn't find any significant appetite to run (or pay someone to run) an alternative that might replace it.
